Add MotorOutputSim

haydenheroux opened the pull request:
The purpose of this pull request is to add simulation support for motor driven mechanisms using system identification constants. This class will adapt WPILib's motor simulation to our **MotorOutput** interface. This allows for substituting any **MotorOutput** with a simulated system. Ideally, a **MotorOutput** should perfectly match the behavior of a real system if it uses the real system's identified constants. Using a feedforward with the same constants should cause the intended dynamics in the simulation, since the feedforward is the inverse of the physics simulation.

haydenheroux committed:
Add MotorOutputSim.

haydenheroux commented:
This design is flawed because **LinearSystemId**  can identify many types of systems, including arms and elevators, and this class could be able to handle all of them. A possible change would be to only have a constructor for **LinearSystem**. This would permit these:
- **new MotorOutputSim(LinearSystemId.createFlywheelSystem(...))**
- **new MotorOutputSim(LinearSystemId.createElevatorSystem(...))**
- **new MotorOutputSim(LinearSystemId.createSingleJointedArmSystem(...))**

Each of these constructors allows passing in parameters from CAD (moment of inertia, drum size, length, etc.), while also permitting **new MotorOutputSim(LinearSystemId.identifyPositionSystem(kV, kA))**.

Adding  custom voltage offsets (e.g. for simulating the effects of gravity) would be a challenging addition.

haydenheroux committed:
Change constructor.

haydenheroux commented:
Consider this class an adapter from WPILib's **DCMotorSim** to our **MotorOutput** interface. The current implementation simulates voltage loss due to static friction. Future work would include separating these concerns by taking the **DCMotorSim** by composition rather than inheritance in an explicit **DCMotorSimOutput** class. Then, voltage losses could be simulated modifying the **setVoltage** method in an extending class. Simulating the voltage loss due to gravity for an arm is still awkward because the angular position of the motor as reported by the sim may need to be translated to the angular position of the arm for calculation.

lddtwo456 commented:
Since each type of sim (FlywheelSim, ElevatorSim, SingleJointedArmSim) is interfaced with differently, makes use of kG and kS differently, and would have things like stator and supply current calculated different, for now would it not be easier to just make multiple implementations for MotorOutput (MotorOutputFlywheelSim, MotorOutputElevatorSim, MotorOutputArmSim), and combine them later when we can see more easily where there are crossovers and what can be done to simplify it all into one implementation MotorOutputSim?

Also, I didn't know about LinearSystemId.identifyPositionSystem(kV, kA) and LinearSystemId.identifyVelocitySystem(kV, kA), use of those might mean we could have the sim constructors take only kV kA kS kG as parameters, making simulating our real systems a bit easier either through using real gains from real tuning, or ballparking estimated gains. If we have seperate implementations of MotorOutput for each type of sim, we could even have two different constructors, one from physical constants we can get from CAD, and one from gains we can get from real testing.

lddtwo456 commented:
oh I did not see commits I thought this was an issue I'll review fr

lddtwo456 commented:
Hm ok I like the idea of this, I didn't think about extending DCMotorSim itself that's smart. I don't really see how kG in an arm can be simulated by a LinearSystem if the constructor for that linear system doesn't take any info about the center of mass of the system though?  Is it liiiike... assuming the initial position of the system is 0, with the line from the pivot to the center of mass parallel to the ground, and somehow doing some calc that I don't want to do to find that you can find voltage to overcome gravity at any angle given only the moment of inertia of the arm and the assumption I mentioned earlier? But that can't be true, because the moment of inertia of a flywheel is not 0, but kG is 0. Maybe this becomes doable as long as the center of mass is not concentric eith the pivot?

haydenheroux committed:
Demo of elevator simulation using MotorOutputSim.

haydenheroux committed:
Report applied voltage as motor voltage.

This fixes a bug where the motor voltage was reported as the effective voltage, instead of the actual applied voltage. This may motivate examination of if the system's electrical properties are correctly simulated.

haydenheroux committed:
Arm linearization with MotorOutputSim.

haydenheroux committed:
Arm linearization with MotorOutputSim.

haydenheroux committed:
Improve MotorOutputSim comments.

haydenheroux committed:
Restore RobotContainer.

lddtwo456 commented:
Ah I see I had assumed it was alrady in a finished state when I commented and there was no kG yet, it makes sense how it is now. Could we make the constructor that takes a **Supplier<Voltage<what>>** for kG instead take a **Function<Angle, Voltage>**, so that the voltage output could actually vary depending on arm angle? Then in the **MotorOutputSim** class there could be a variable **Supplier<Voltage<huh>> kG**, and in the constructor with **Function<Angle, Voltage> kG** as an input, you could do `this.kG = () -> kG.apply(Radians.of(sim.getAngularPositionRad()));`, and for the constructor with constant kG, do `this.kG = () -> kG;`

I guess the same thing could also be achieved just by making the supplier input someting like `() -> Math.cos(motorValues.position) * kG;`, but the constructor for **MotorOutputSim** is being called in a factory and won't have access to motor values.

haydenheroux commented:
https://github.com/user-attachments/assets/52e56a0f-dc1a-4c4e-91aa-cce1d7700291

```java
  public Command getAutonomousCommand() {
      var dt = Seconds.of(0.02);
      var gearbox = DCMotor.getKrakenX60(2);
      // Higher values represent higher maximum velocities
      var kV = Volts.of(0.1);
      // Higher values penalize acceleration (heavier mechanisms, higher moments of inertia, etc.)
      var kA = Volts.of(0.025);
      // Assume no static friction, to simplify feedforward calculation
      var kS = Volts.of(0);
      // When the arm is perpendicular to the ground, this much voltage is required to hold it up
      var kG = Volts.of(0.575);
      Function<Angle, Voltage> calculateKg = position -> kG.times(Math.cos(position.in(Radians)));
      var sim = new DCMotorSim(LinearSystemId.identifyPositionSystem(kV.in(Volts), kA.in(Volts)), gearbox);
      // Initialize arm in down position
      var initialAngle = Degrees.of(-90);
      sim.setAngle(initialAngle.in(Radians));
      var output = new MotorOutputSim(sim, kS, calculateKg);
      var values = new MotorValues();
      // Goal is to stabilize arm in up position (unstable)
      var goal = Degrees.of(90);
      var kP = Volts.per(Radian).of(0.25);
      // Render the mechanism for some nice feedback
      var mechanism = new Mechanism2d(3, 3);
      var root = mechanism.getRoot("demo", 1.5, 1.5);
      var wrist = root.append(new MechanismLigament2d("arm", 1.0, initialAngle.in(Degrees), 6, new Color8Bit(Color.kLimeGreen)));
      SmartDashboard.putData("Demo", mechanism);
      Runnable update = () -> {
          output.updateValues(values, dt);
          SmartDashboard.putNumber("Position (deg)", values.position.in(Degrees));
          SmartDashboard.putNumber("Velocity (dps)", values.velocity.in(DegreesPerSecond));
          SmartDashboard.putNumber("Voltage (V)", values.motorVoltage.in(Volts));
          wrist.setAngle(values.position.in(Degrees));
      };
      Runnable controlLoop = () -> {
          var ff = calculateKg.apply(values.position);
          var error = goal.minus(values.position);
          var fb = error.timesRatio(kP);
          output.setVoltage(ff.plus(fb));
      };
      Runnable voidControlLoop = () -> {
          output.setVoltage(Volts.zero());
      };
      var swingUp = Commands.run(() -> {
          update.run();
          controlLoop.run();
      }).withTimeout(8);
      var nudge = Commands.runOnce(() -> {
          sim.setAngle(goal.plus(Degrees.of(0.01)).in(Radians));
          sim.setAngularVelocity(0);
      });
      var fallDown = Commands.run(() -> {
          update.run();
          voidControlLoop.run();
      }).withTimeout(8);
      return Commands.sequence(swingUp, nudge, fallDown).repeatedly();
  }

```

haydenheroux committed:
kG as a function of position.

haydenheroux committed:
Merge branch 'gamma-rewrite' into hayden/kv_ka_position_sim

haydenheroux committed:
Use MotorOutputSim for elevator.

lddtwo456 commented:
lgtm, we will move **setPosition()** into **MotorOutput** in another issue to fix the problem of the **Angle** input to the **Function<Angle, Voltage>** not taking into account position offset and not being able to do so cleanly since offset is not accessible from inside the **MotorOutputSim** class or even the subsystem factory with the way things are right now.
